gen/emit/EmitJulia.cpp

```diff
@@ -548,27 +592,194 @@ struct EmitJulia : public Emitter
+                // extract any existing sub-modules and types contained in the
+                // module. We'll include those entries in the final generated
+                // module file.
+                moduleJlFp = fopen(moduleJlFname.c_str(), "r");
+                if (!moduleJlFp) {
+                    perror("fopen");
+                    return -1;
+                }
+                // Read chunks of the existing file line by line
+                while(!feof(moduleJlFp)) {
+                    char buf[4096];
+                    memset(buf, 0, sizeof(buf));
+                    char *result = fgets(buf, sizeof(buf)-1, moduleJlFp);
+                    if (!result) break;
+
+                    auto words = StringUtil::split(StringUtil::strip(buf), ' ');
+                    if (words.size() >= 2 && words[0] == "import") {
+                        // If this line matches "import foo", then store "foo" in
+                        // the set of imports
+                        moduleJlImports.insert(words[1]);
+                    } else if (words.size() >= 6 &&
+                               words[0] == "@eval" &&
+                               words[2] == "module" &&
+                               words[4] == ";" &&
+                               words[5] == "end") {
+                        // Otherwise, if the line matches:
+                        // @eval foo module bar.baz ; end
+                        // then store foo.bar.baz in the submodules set
+                        moduleJlSubmodules.insert(words[1] + "." + words[3]);
+                    }
+                }
+                fclose(moduleJlFp);
+                moduleJlFp = nullptr;
+            }
+            // Regardless of whether the module existed, we'll create a new file
+            moduleJlFp = fopen(moduleJlFname.c_str(), "w");
+            if (!moduleJlFp) {
+                perror("fopen");
+                return -1;
+            }
+            fprintf(moduleJlFp, "\"\"\"ZCM package %s.jl file\n"
+                     "This file automatically generated by zcm-gen.\n"
+                     "DO NOT MODIFY BY HAND!!!!\n"
+                     "\"\"\"\n\n"
+                     "module %s; end\n\n",
+                     pdname.c_str(), pdname.c_str());
+            for (auto& submod : moduleJlSubmodules) {
+                auto parts = StringUtil::split(submod, '.');
+                if (parts.size() >= 2) {
+                    // Restore each of the submodules we parsed from the
+                    // existing file (if any), and also generate the submodule
+                    // corresponding to the current package (if necessary)
+                    vector<string> parentParts(parts.begin(), parts.end() - 1);
+                    auto parent = StringUtil::join(parentParts, '.');
+                    auto module = parts.at(parts.size() - 1);
+                    fprintf(moduleJlFp, "@eval %s module %s ; end\n", parent.c_str(), module.c_str());
```

Comment:
note that the @eval ... syntax seems to only work with Julia 0.6. If that's not a backwards compatible operation, I'm ok with dropping 0.5 support, but then we should also upgrade the rest of the Julia support in zcm to use 0.6 syntax

Reply:
We can replace:

```
@eval Foo module Bar
end
```

with:

```
eval(Foo, :(module Bar
end))
```
which works on v0.5 and v0.6